docs/source/plugin/advanced.rst

```diff
@@ -64,3 +64,115 @@ a QUIT/PART event, and they will be available after a JOIN event.
+
+Declaring requests: the ``capability`` decorator
+------------------------------------------------
+
+In :mod:`sopel.plugin` there is an advanced :class:``capability`` decorator: it
+is a class that acts like a decorator, and can be used with or without a
+callback to decorate::
+
+    # plugin file
+    from sopel import plugin
+
+    # this will register a capability request
+    CAP_ACCOUNT_TAG = plugin.capability('account-tag')
+
+    # this will work as well
+    @plugin.capability('message-prefix')
+    def cap_message_prefix(cap_req, bot, acknowledged):
+        # do something if message-prefix is ACK or NAK
+        ...
+
+.. autoclass:: sopel.plugin.capability
+   :members:
+
+.. autoclass:: sopel.plugin.CapabilityNegotiation
+   :members:
+
+
+Working with capabilities
+-------------------------
+
+A plugin that requires capabilities, or that can enhance its features with
+capabilities, should rely on the bot's
+:attr:`capability manager <sopel.irc.AbstractBot.capabilities>`::
+
+    @plugin.command('mycommand')
+    def mycommand_handler(bot, trigger):
+        if bot.capabilities.is_enabled('cap1'):
+            # be fancy with enhanced capabilities
+        else:
+            # stick to the basics
+
+The manager's :meth:`~sopel.irc.capabilities.Capabilities.is_enabled` in
+particular is the most interesting, as it allows a plugin to always know if a
+capability is available or not.
+
+Ending negociations
+-------------------
+
+Sopel automatically sends a ``CAP END`` message when all requests are handled.
+However in some case, a plugin author may need to delay the end of the
+negociation to perform an action that must be done first.
+
+In that case, plugin author must return
+:attr:`~sopel.plugin.CapabilityNegotiation.CONTINUE` in their callback.
+
+This is the case for the SASL authentification, as seen in the ``coretasks``
+internal plugin that manages that::
+
+    @plugin.capability('sasl')
+    def cap_sasl_handler(cap_req, bot, acknowledged):
+        # ... <skip for readability> ...
+        bot.write(('AUTHENTICATE', mech))
+
+        # If we want to do SASL, we have to wait before we can send CAP END.
+        # So if we are, wait on 903 (SASL successful) to send it.
+        return plugin.CapabilityNegotiation.CONTINUE
+
+Later on, the plugin uses the
+:meth:`~sopel.bot.Sopel.resume_capability_negotiation` method to tell the bot
+that the request is complete, and the bot will send the ``CAP END``
+automatically::
+
+    @plugin.event(events.RPL_SASLSUCCESS)
+    @plugin.thread(False)
+    @plugin.unblockable
+    @plugin.priority('medium')
+    def sasl_success(bot: SopelWrapper, trigger: Trigger):
+        """Resume capability negotiation on successful SASL auth."""
+        LOGGER.info("Successful SASL Auth.")
+        bot.resume_capability_negotiation(CAP_SASL.cap_req, 'coretasks')
+
+.. important::
+
+    Note that these plugin callable should always use ``@plugin.thread(False)``
```

Comment:
```suggestion
    Plugin callables that modify the bot's capability negotiation state should always use ``@plugin.thread(False)``
```

No rewrap because I'm lazy and the PR already has a bunch of stuff that needs squashing down anyway.